fdbserver/StorageMetrics.actor.h

```diff
@@ -410,6 +410,104 @@ struct StorageServerMetrics {
+		// shard.end.printable().c_str());
+		while (endKey != byteSample.sample.end()) {
+			// printf("+++++++++++++ Looking at range (%s, %s)\n",beginKey.printable().c_str(),
+			// (*endKey).printable().c_str());
+			if (*endKey > shard.end) endKey = byteSample.sample.lower_bound(shard.end);
+			if (*endKey == beginKey) {
+				chunkSize += baseChunkSize;
+				endKey = byteSample.sample.index(byteSample.sample.sumTo(byteSample.sample.lower_bound(beginKey)) +
+				                                 chunkSize);
+				continue;
+			}
+			// printf("+++++++++++++ This range has size %d and read sample size %d\n",
+			// byteSample.getEstimate(KeyRangeRef(beginKey, *endKey)), bytesReadSample.getEstimate(KeyRangeRef(beginKey,
+			// *endKey)));
+			if (bytesReadSample.getEstimate(KeyRangeRef(beginKey, *endKey)) / chunkSize > readDensityRatio) {
+				auto range = KeyRangeRef(beginKey, *endKey);
+				if (!toReturn.empty() && toReturn.back().end == range.begin) {
+					auto updatedTail =
+					    KeyRangeRef(toReturn.back().begin,
+					                *endKey); // in case two consecutive chunks both are over the ratio, merge them.
+					toReturn.pop_back();
+					toReturn.push_back(updatedTail);
+					// TraceEvent("RHDLogRangeReturned")
+					//     .detail("ShardStart", range.begin.printable().c_str())
+					//     .detail("ShardEnd", range.end.printable().c_str())
+					//     .detail("ReadBandWith", bytesReadSample.getEstimate(range))
+					//     .detail("Ratio", bytesReadSample.getEstimate(range) / chunkSize)
+					//     .detail("RatioThrethold", readDensityRatio);
+				} else {
+					// TraceEvent("RHDLogRangeReturned")
+					//     .detail("ShardStart", range.begin.printable().c_str())
+					//     .detail("ShardEnd", range.end.printable().c_str())
+					//     .detail("ReadBandWith", bytesReadSample.getEstimate(range))
+					//     .detail("Ratio", bytesReadSample.getEstimate(range) / chunkSize)
+					//     .detail("RatioThrethold", readDensityRatio);
+					toReturn.push_back(range);
+				}
+			}
+			beginKey = *endKey;
+			chunkSize = baseChunkSize;
+			endKey =
+			    byteSample.sample.index(byteSample.sample.sumTo(byteSample.sample.lower_bound(beginKey)) + chunkSize);
+		}
+		// printf("Hot range vector contains [%d] ranges.\n", toReturn.size());
+		// for (auto i = toReturn.begin(); i < toReturn.end(); i++) {
+		// printf("Range [%s, %s) is a read dense range.\n", i->begin.printable().c_str(), i->end.printable().c_str());
+		// }
+		return toReturn;
+	}
+
+	void getReadHotRanges(ReadHotSubRangeRequest req) {
+		ReadHotSubRangeReply reply;
+		std::vector<KeyRangeRef> v = getReadHotRanges(req.keys, SERVER_KNOBS->SHARD_MAX_READ_DENSITY_RATIO,
+		                                              SERVER_KNOBS->READ_HOT_SUB_RANGE_CHUNK_SIZE,
+		                                              SERVER_KNOBS->SHARD_READ_HOT_BANDWITH_MIN_PER_KSECONDS);
+		reply.readHotRanges = VectorRef<KeyRangeRef>(v.begin().base(), v.size());
```

Comment:
This is an error from one of the windows builds:

```
 20>c:\users\fdb\jenkins\foundationdb-ci.foundationdb.org\workspace\jobs\prb-windows\fdbserver\storagemetrics.actor.h(507): error C2039: 'base': is not a member of 'std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>' [C:\Users\fdb\jenkins\foundationdb-ci.foundationdb.org\workspace\jobs\prb-windows\fdbserver\fdbserver.vcxproj]
                 with
                 [
                     _Ty=KeyRangeRef
                 ] (compiling source file Resolver.actor.g.cpp)
         c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.16.27023\include\vector(416): note: see declaration of 'std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>'
                 with
                 [
                     _Ty=KeyRangeRef
                 ] (compiling source file Resolver.actor.g.cpp)
    20>c:\users\fdb\jenkins\foundationdb-ci.foundationdb.org\workspace\jobs\prb-windows\fdbserver\storagemetrics.actor.h(507): error C2440: '<function-style-cast>': cannot convert from 'unsigned __int64' to 'VectorRef<KeyRangeRef,VecSerStrategy::FlatBuffers>' (compiling source file Resolver.actor.g.cpp) [C:\Users\fdb\jenkins\foundationdb-ci.foundationdb.org\workspace\jobs\prb-windows\fdbserver\fdbserver.vcxproj]
         c:\users\fdb\jenkins\foundationdb-ci.foundationdb.org\workspace\jobs\prb-windows\fdbserver\storagemetrics.actor.h(507): note: No constructor could take the source type, or constructor overload resolution was ambiguous (compiling source file Resolver.actor.g.cpp)
```

Reply:
Ha. Let me fix that.